### PR TITLE
Experimental AVR target improvements

### PR DESCRIPTION
**Pull request by gergoerdi**
Noob alert: **Please review very carefully** -- I haven't even looked at LLVM's internals until a week ago.

The commits should be fairly self-explanatory; the fixes are for the following issues:

* https://github.com/avr-rust/rust/issues/44
* https://github.com/avr-rust/rust/issues/46
* https://github.com/avr-rust/rust/issues/47

### CODE REVIEW

**Comment by shepmaster** [comment]
Thanks for the patches! I think our preference is to land everything in upstream LLVM before it finds its way into this repository, but this seems like a good place to track those patches *before* then (and also to allow @dylanmckay to grab them and review them)

Said another way: I think this PR can stay open, but not be merged until each of the commits has been merged into LLVM.

**Comment by gergoerdi** [comment]
Oh, I thought it was exactly the other way round -- that this repo is to stage dubious changes until they can be upstreamed to LLVM.

BTW, most definitely don't merge this PR, since the last three commits break structure unpacking: https://github.com/avr-rust/rust/issues/48

**Comment by gergoerdi** [comment]
This branch now has a ton of useful fixes with which I was able to compile a non-trivial Rust app to AVR (I hope to write a blog post about it shortly). Can I leave it in your hands to upstream it to LLVM proper?

**Comment by shepmaster** [comment]
@gergoerdi each separate fix to LLVM should have one or more tests associated with it; otherwise it's highly likely that any fixes will regress with time. 

@dylanmckay any other things you'd like to see?

File: lib/Target/AVR/AVRInstrInfo.td [hunk 1]
```diff
@@ -2057,6 +2109,10 @@ def : Pat<(store i8:$src, (i16 (AVRWrapper tglobaladdr:$dst))),
 def : Pat<(store i16:$src, (i16 (AVRWrapper tglobaladdr:$dst))),
           (STSWKRr tglobaladdr:$dst, i16:$src)>;
 
+// Temporary workaround -- this one is BOGUS!
```
Comment (dylanmckay) [inline lib/Target/AVR/AVRInstrInfo.td:2111]: Why is this temporary? What is it working around? Can you pop a comment somewhere with an explanation?

**Comment by gergoerdi** [comment]
@shepmaster OK I'll try to find the time to add tests for the new fixes. My question still stands though -- can you or @dylanmckay take care of the upstreaming? I'm not yet invested enough in this to want to take on that hassle.

**Comment by gergoerdi** [comment]
Also, I think there's three categories of my commits here:

1. Definite bugfixes, no reason not to merge them:

- [x] https://github.com/avr-rust/llvm/pull/1/commits/c7d5217f866a3e07bb9dd2b337b61463356dea77 (avr-rust/rust#49)
- [x] https://github.com/avr-rust/llvm/pull/1/commits/9fe62d7238036defc258118a2f2b74127fb771f1 (avr-rust/rust#50)
- [x] https://github.com/avr-rust/llvm/pull/1/commits/910429cca6d7b1bc9cd0d2f22f1bd1d011c54fee (avr-rust/rust#52)

2. Definite improvement, but implementation could be better

- [ ] https://github.com/avr-rust/llvm/pull/1/commits/383e429f720ab50122bb7adcd22aabcc1e361671 (avr-rust/rust#47)

3. Desperate struggling piling hacks upon hacks fighting the Harvard architecture

- [ ] https://github.com/avr-rust/llvm/pull/1/commits/de2539285a6ab0287466cbe9a79a8757dba964f3 (avr-rust/rust#52)
- [ ] https://github.com/avr-rust/llvm/pull/1/commits/07a96e39d25cd2db8b710a826e941bb5aa7b0a16 (avr-rust/rust#52)
- [ ] https://github.com/avr-rust/llvm/pull/1/commits/bff3af337679c389c755b53ab826f2816f18dae4 (avr-rust/rust#52)
- [ ] https://github.com/avr-rust/llvm/pull/1/commits/65f3f2abb64c23fe29aacdd602d5346d388945b1  (avr-rust/rust#52)

I don't think commits from the third category should be merged; I'll open a separate ticket where we can hopefully come up with a proper solution to `PROGMEM`-related issues.

**Comment by dylanmckay** [comment]
I'm happy to take on upstreaming @gergoerdi, will keep you updated.

**Comment by dylanmckay** [comment]
I'm going to put this comment here to track which commits I have upstreamed

Commits upstreamed

* [AVR] Add pseudo-instructions for relaxed branches.

**Comment by dylanmckay** [comment]
It looks like there's some stuff here that still needs to be merged

I will change the base branch to see more clearly what is missing that should go into master

**Comment by dylanmckay** [comment]
That didn't help, the histories are different!

**Comment by dylanmckay** [comment]
Stuff to merge

* "[AVR] Elaborate LDWRdPtr into `ld r, X++; ld r+1, X`"
* Grab `mayLoad` change from the diff out of "[AVR] Add ISel pattern which maches load-from-global, compile it to LPMRdZ"
* "[AVR] Implement LPMWRdZ pseudo-instruction's expansion."

I believe everything else is already upstreamed, or were temporary hacks

**Comment by dylanmckay** [comment]
Ran tests locally, all pass

**Comment by dylanmckay** [comment]
Wrong merge request! did not mean to merge, nor did I run tests for this PR.

Reopening now, have reverted

**Comment by dylanmckay** [comment]
Looks like there's no reopen button, will just cherry-pick stuff now

**Comment by dylanmckay** [comment]
Upstreamed ""[AVR] Elaborate LDWRdPtr into ld r, X++; ld r+1, X"" in [r314896](http://llvm.org/viewvc/llvm-project?view=revision&revision=314896).

I also fixed a bunch of tests before committing

**Comment by dylanmckay** [comment]
Upsteamed `mayLoad` factoring out in [r314897](http://llvm.org/viewvc/llvm-project?view=revision&revision=314897)

**Comment by dylanmckay** [comment]
Upsteamed "[AVR] Implement LPMWRdZ pseudo-instruction's expansion." in [r314898](http://llvm.org/viewvc/llvm-project?view=revision&reision=314898)

**Comment by dylanmckay** [comment]
Cherry-picked the upstreamed commits to avr-rust/llvm in

06d0f6b412b [AVR] Implement LPMWRdZ pseudo-instruction's expansion.
f6d00a7140b [AVR] Factor out mayLoad in tablegen patterns
d4274bd15fc [AVR] Elaborate LDWRdPtr into `ld r, X++; ld r+1, X`

**Comment by dylanmckay** [comment]
Updated version of LLVM in for Rust in https://github.com/avr-rust/rust/commit/a22dd4652b512598bb2783740abc70d98296a16f

**Comment by dylanmckay** [comment]
Added these commits to the current AVR 5.0.1 merge request [PR34829](https://bugs.llvm.org/show_bug.cgi?id=34829)

**Comment by dylanmckay** [comment]
Have double checked all the commits in this PR and everything looks good.

Thanks for the patches @gergoerdi, sorry for taking so long, completely forgot about this PR!

**Pull request merged**